python/lsst/resources/gs.py

```diff
@@ -0,0 +1,279 @@
+                    log.debug("Copied %d bytes out of %d (%s to %s)", bytes_copied, total_bytes, src, self)
+                    if rewrite_token is None:
+                        # Copy has completed
+                        break
+        else:
+            # Use local file and upload it
+            with src.as_local() as local_uri:
+                with time_this(log, msg=timer_msg, args=timer_args):
+                    self.blob.upload_from_filename(local_uri.ospath)
+
+        # This was an explicit move requested from a remote resource
+        # try to remove that resource
+        if transfer == "move":
+            # Transactions do not work here
+            src.remove()
+
+    @contextlib.contextmanager
+    def open(
+        self,
+        mode: str = "r",
+        *,
+        encoding: Optional[str] = None,
+        prefer_file_temporary: bool = False,
+    ) -> Iterator[IO]:
+        # Docstring inherited
+        if "x" in mode:
+            if self.exists():
+                raise FileExistsError(f"File at {self} already exists.")
+            mode = mode.replace("x", "w")
+        # The GCS API does not support append or read/write modes so for
+        # those we use the base class implementation.
+        if "+" in mode or "a" in mode:
+            with super().open(mode, encoding=encoding, prefer_file_temporary=prefer_file_temporary) as buffer:
+                yield buffer
+        else:
+            with self.blob.open(mode, encoding=encoding) as buffer:
+                yield buffer
+            if "w" in mode and self.exists():
+                # Force the blob to be resynced.
+                self._blob = None
+
+    def walk(
+        self, file_filter: Optional[Union[str, re.Pattern]] = None
+    ) -> Iterator[Union[List, Tuple[ResourcePath, List[str], List[str]]]]:
+        # We pretend that GCS uses directories and files and not simply keys.
+        if not (self.isdir() or self.is_root):
+            raise ValueError(f"Can not walk a non-directory URI: {self}")
+
+        if isinstance(file_filter, str):
+            file_filter = re.compile(file_filter)
+
+        # Limit each query to a single "directory" to match os.walk
+        # We could download all keys at once with no delimiter and work
+        # it out locally but this could potentially lead to large memory
+        # usage for millions of keys. It will also make the initial call
+        # to this method potentially very slow. If making this method look
+        # like os.walk was not required, we could query all keys with
+        # pagination and return them in groups of 1000, but that would
+        # be a different interface since we can't guarantee we would get
+        # them all grouped properly across the 1000 limit boundary.
+        prefix = self.relativeToPathRoot if not self.is_root else ""
+        prefix_len = len(prefix)
+        dirnames: Set[str] = set()
```

Comment:
S3 has a simple list; why do these differ?

Reply:
I don't suppose it's possible for duplicates so this can be a list.

Reply:
Ah. It was because I was concerned that it's possible for the same prefix to turn up in multiple pages and did not want to duplicate them.